README.md
# ActiveRecord::PGEnum [![Build Status](https://circleci.com/gh/alassek/activerecord-pg_enum.svg?style=shield)](https://app.circleci.com/pipelines/github/alassek/activerecord-pg_enum)

**Note**: This was originally written before Rails added support for native enums. See the comments in [version
support](#version-support).

The `enum` feature in Rails has bad developer ergonomics. It uses integer types at the DB layer, which means trying to understand SQL output is a pain.

Using the easy form of the helper syntax is a minor footgun:

```ruby
enum status: %w[new active archived]
```

It's not obvious that the above code is order-dependent, but if you decide to add a new enum anywhere but the end, you're in trouble.

If you choose the use `varchar` fields instead, now you have to write annoying check constraints and lose the efficient storage.

```ruby
enum status: { new: "new", active: "active", archived: "archived" }
```

Nobody has time to write that nonsense.

## Enumerated Types: The Best of Both Worlds

Did you know you can define your own types in PostgreSQL? You can, and this type system also supports enumeration.

```SQL
CREATE TYPE status_type AS ENUM ('new', 'active', 'archived');
```

Not only does this give you full type safety at the DB layer, the implementation is highly efficient. An enum value only takes up [four bytes](https://www.postgresql.org/docs/11/datatype-enum.html).

The best part is that PostgreSQL supports inserting new values at any point of the list without having to migrate your data.

```SQL
ALTER TYPE status_type ADD VALUE 'pending' BEFORE 'active';
```

## schema.rb Support

The principle motivation of this gem is to seamlessly integrate PG enums into your `schema.rb` file. This means you can use them in your database columns without switching to `structure.sql`.

```ruby
ActiveRecord::Schema.define(version: 2019_06_19_214914) do

  # These are extensions that must be enabled in order to support this database
  enable_extension "plpgsql"

  create_enum "status_type", %w[new pending active archived]
  
  create_table "orders", id: :serial, force: :cascade do |t|
    t.enum "status", as: "status_type", default: "new"
  end

end
```

## Version support

Every version of Rails with an `enum` macro is supported. This means 4.1 through master. Yes, this was annoying and difficult.

The monkeypatches in this library are extremely narrow and contained; the dirty hacks I had to do to make 4.1 work, for instance, have no impact on 6.0.

Monkeypatching Rails internals is **scary**. So this library has a comprehensive test suite that runs against every known minor version.

Rails 7 added support for native enums, but they have so far neglected to support altering or dropping enums in the API,
so this gem remains to fill in the gaps. I expect to slowly deprecate this over time.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'activerecord-pg_enum'
```

And then execute:

    $ bundle

## Usage

### Migrations

Defining a new ENUM

```ruby
class AddContactMethodType < ActiveRecord::Migration[5.2]
  def up
    create_enum "contact_method_type", %w[Email Phone]
  end

  def down
    drop_enum "contact_method_type"
  end
end
```

Adding a value to an existing ENUM (you must disable the wrapping transaction on PostgreSQL versions older than 12)

```ruby
class AddSMSToContactMethodType < ActiveRecord::Migration[5.2]
  disable_ddl_transaction!

  def up
    add_enum_value "contact_method_type", "SMS", before: "Phone"
  end

  def down
    raise ActiveRecord::IrreversibleMigration
  end
end
```

Adding an enum column to a table

```ruby
class AddStatusToOrder < ActiveRecord::Migration[5.2]
  def change
    change_table :orders do |t|
      t.enum :status, as: "status_type", default: "new"
    end
  end
end
```

Renaming an enum type

```ruby
class RenameStatusType < ActiveRecord::Migration[6.0]
  def change
    rename_enum "status_type", to: "order_status_type"
  end
end
```

```SQL
ALTER TYPE status_type RENAME TO order_status_type;
```

**PostgreSQL 10+ required**:

Changing an enum label

```ruby
class ChangeStatusHoldLabel < ActiveRecord::Migration[6.0]
  def change
    rename_enum_value "status_type", from: "on hold", to: "OnHold"
  end
end
```

```SQL
ALTER TYPE status_type RENAME VALUE 'on hold' TO 'OnHold';
```

### Module Builder

```ruby
class ContactInfo < ActiveRecord::Base
  include PGEnum(contact_method: %w[Email SMS Phone])
end
```

The generated module calls the official `enum` method converting array syntax into strings. The above example is equivalent to:

```ruby
class ContactInfo < ActiveRecord::Base
  enum contact_method: { Email: "Email", SMS: "SMS", Phone: "Phone" }
end
```

Additionally, `enum` options are fully supported, for example
```ruby
class User < ActiveRecord::Base
  include PGEnum(status: %w[active inactive deleted], _prefix: 'user', _suffix: true)
end
```

is equivalent to
```ruby
class User < ActiveRecord::Base
  enum status: { active: 'active', inactive: 'inactive', deleted: 'deleted' }, _prefix: 'user', _suffix: true
end
```

There's no technical reason why you couldn't detect enum columns at startup time and automatically do this wireup, but I feel that the benefit of self-documenting outweighs the convenience.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `appraisal rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

Test a specific version with `appraisal 6.0 rake spec`. This is usually necessary because different versions have different Ruby version support.

To install this gem onto your local machine, run `bundle exec rake install`.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/alassek/activerecord-pg_enum.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2022-07-17
### Added
- Support for 7.0

### Changed
- ActiveRecord added `enum_types` interface so monkeypatch has been removed. **Note:** output format has changed.
- `create_enum` uses framework interface, replacing `:as` kwarg with `:enum_type`

## [1.2.2] - 2021-01-07
### Fixed
- Don't mutate `Gem::Version#canonical_segments` (@agrobbin)

## [1.2.1] - 2021-01-05
### Fixed
- Argument bug that surfaced in Ruby 3.0

## [1.2.0] - 2020-12-09
### Added
- Support for 6.1

## [1.1.0] - 2020-01-18
### Added
- `rename_enum` command for changing the type name
- `rename_enum_value` command for changing an enum label (PostgreSQL 10+ only)

### Changed
- Refactored some code to eliminate deprecation warnings in 2.7 related to kwargs

## [1.0.5] - 2019-11-22
### Fixed
- Don't include schema name in dumped enum types to be consistent with the way tables are dumped (@TylerRick)

## [1.0.4] - 2019-11-16
### Fixed
- Dump enums declared in non-public schemas
- Fix missing `enum` method with using `change_table`

## [1.0.3] - 2019-10-13
### Fixed
- Allow enum labels to contain spaces (@AndrewSpeed)

## [1.0.2] - 2019-08-29
- Move the active_record load hook to a different place to ensure things run in the correct order

## [1.0.1] - 2019-08-16
- Update Rails 6 support to 6.0-final

## [1.0.0] - 2019-06-23
### Added
- Support for 4.1 and 4.2

### Changed
- Moved module builder to top-level `PGEnum()` method

## [0.4.0] - 2019-06-19
### Added
- `enum` method on `TableDefinition`

### Changed
- Refactored init hook to be much more flexible
- Removed `ActiveRecord::PGEnum::Helper` in favor of `ActiveRecord::PGEnum()` module builder

## [0.3.0] - 2019-03-03
- Support for 5.0 and 5.1
- Change travis config to test against oldest supported version of ruby

## [0.2.1] - 2019-02-22
- Fixed a bug in the `SchemaDumper` output

## [0.2.0] - 2018-08-18
### Changed
- Change API of `create_enum` to take two arguments instead, the name and an array of values

## [0.1.1] - 2018-08-14
- Bump Ruby requirement to 2.2.2 (for Rails 5.2) until earlier framework versions are supported.
- `enum_types` are listed in alphabetical order

## [0.1.0] - 2018-08-12
### Added
- `add_to_enum` for modifying existing types
- `ActiveRecord::Migration::CommandRecorder` is patched to make `create_enum`, and `drop_enum` reversible.

### Changed
- Ruby 2.1 is defined as the earliest supported version (for kwargs)
- `ActiveRecord::ConnectionAdapters::PostgreSQLAdapter#enum_types` returns a Hash instead of a nested Array


lib/active_record/pg_enum.rb
require "active_record"
require "active_support/lazy_load_hooks"

module ActiveRecord
  module PGEnum
    KNOWN_VERSIONS = %w[4.1 4.2 5.0 5.1 5.2 6.0 6.1 7.0].map { |v| Gem::Version.new(v) }

    class << self
      attr_reader :enabled_version

      def install(version)
        @enabled_version = approximate_version(version)

        # Don't immediately fail if we don't yet support the current version.
        # There's at least a chance it could work.
        if !KNOWN_VERSIONS.include?(enabled_version) && enabled_version > KNOWN_VERSIONS.last
          @enabled_version = KNOWN_VERSIONS.last
          warn "[PGEnum] Current ActiveRecord version unsupported! Falling back to: #{enabled_version}"
        end

        initialize!
      end

      def register(patch, &block)
        monkeypatches[patch] = block
      end

      def detected_version
        approximate_version Gem.loaded_specs["activerecord"].version
      end

      private

      def monkeypatches
        @patches ||= {}
      end

      def initialize!
        require "active_record/pg_enum/command_recorder"
        require "active_record/pg_enum/schema_statements"

        Dir[File.join(__dir__, "pg_enum", enabled_version.to_s, "*.rb")].each { |file| require file }
        monkeypatches.keys.each { |patch| monkeypatches.delete(patch).call }
      end

      def approximate_version(version)
        segments = version.respond_to?(:canonical_segments) ? version.canonical_segments.dup : version.segments

        segments.pop     while segments.any? { |s| String === s }
        segments.pop     while segments.size > 2
        segments.push(0) while segments.size < 2

        Gem::Version.new segments.join(".")
      end
    end
  end
end

ActiveSupport.on_load(:active_record) do
  ActiveRecord::PGEnum.install Gem.loaded_specs["activerecord"].version
end

# Declare an enum attribute where the values map to strings enforced by PostgreSQL's
# enumerated types.
#
#   class Conversation < ActiveRecord::Base
#     include PGEnum(status: %i[active archived])
#   end
#
# This is merely a wrapper over traditional enum syntax so that you can define
# string-based enums with an array of values.
def PGEnum(**definitions)
  values = definitions.values.map do |value|
    if value.is_a? Array
      keys   = value.map(&:to_sym)
      values = value.map(&:to_s)

      Hash[keys.zip(values)]
    else
      value
    end
  end

  Module.new do
    define_singleton_method(:inspect) { %{ActiveRecord::PGEnum(#{definitions.keys.join(" ")})} }

    define_singleton_method :included do |klass|
      klass.enum Hash[definitions.keys.zip(values)]
    end
  end
end
